WalletWasabi/WabiSabi/Backend/Rounds/Arena.cs

```diff
@@ -391,39 +395,68 @@ public async Task<ConnectionConfirmationResponse> ConfirmConnectionAsync(Connect
 					throw new WabiSabiProtocolException(WabiSabiProtocolErrorCode.IncorrectRequestedAmountCredentials, $"Round ({request.RoundId}): Incorrect requested amount credentials.");
 				}
 
-				var commitAmountZeroCredentialResponse = round.AmountCredentialIssuer.PrepareResponse(request.ZeroAmountCredentialRequests);
-				var commitVsizeZeroCredentialResponse = round.VsizeCredentialIssuer.PrepareResponse(request.ZeroVsizeCredentialRequests);
-
-				if (round.Phase == Phase.InputRegistration)
+				if (round.Phase is not (Phase.InputRegistration or Phase.ConnectionConfirmation))
 				{
-					alice.SetDeadlineRelativeTo(round.ConnectionConfirmationTimeout);
-					return new(
-						commitAmountZeroCredentialResponse.Commit(),
-						commitVsizeZeroCredentialResponse.Commit());
+					throw new WabiSabiProtocolException(WabiSabiProtocolErrorCode.WrongPhase, $"Round ({request.RoundId}): Wrong phase ({round.Phase}).");
 				}
-				else if (round.Phase == Phase.ConnectionConfirmation)
-				{
-					var state = round.Assert<ConstructionState>();
-
-					// Ensure the input can be added to the CoinJoin
-					state = state.AddInput(alice.Coin);
+			}
 
-					var commitAmountRealCredentialResponse = round.AmountCredentialIssuer.PrepareResponse(realAmountCredentialRequests);
-					var commitVsizeRealCredentialResponse = round.VsizeCredentialIssuer.PrepareResponse(realVsizeCredentialRequests);
+			var amountZeroTask = Task.Run(() => round.AmountCredentialIssuer.PrepareResponse(request.ZeroAmountCredentialRequests), cancellationToken);
+			var vsizeZeroTask = Task.Run(() => round.VsizeCredentialIssuer.PrepareResponse(request.ZeroVsizeCredentialRequests), cancellationToken);
+			Task<ICommitableCredentialsResponse>? amountRealTask = null;
+			Task<ICommitableCredentialsResponse>? vsizeRealTask = null;
 
-					// update state
-					alice.ConfirmedConnection = true;
-					round.CoinjoinState = state;
+			if (round.Phase is Phase.ConnectionConfirmation)
+			{
+				amountRealTask = Task.Run(() => round.AmountCredentialIssuer.PrepareResponse(realAmountCredentialRequests), cancellationToken);
+				vsizeRealTask = Task.Run(() => round.VsizeCredentialIssuer.PrepareResponse(realVsizeCredentialRequests), cancellationToken);
+			}
```

Comment:
Does it make sense to do it here? I am not sure if I am missing something but why not to do it in the  `switch case` ?

Reply:
The idea here is to precalculate, meanwhile, we are waiting for the lock in line 420.